virtcontainers/device/drivers/generic.go

```diff
@@ -27,38 +32,98 @@ func NewGenericDevice(devInfo *config.DeviceInfo) *GenericDevice {
 
 // Attach is standard interface of api.Device
 func (device *GenericDevice) Attach(devReceiver api.DeviceReceiver) error {
-	if device.DeviceInfo.Hotplugged {
+	skip, err := device.bumpAttachCount(true)
+	if err != nil {
+		return err
+	}
+	if skip {
 		return nil
 	}
-
+	device.AttachCount = 1
 	return nil
 }
 
 // Detach is standard interface of api.Device
 func (device *GenericDevice) Detach(devReceiver api.DeviceReceiver) error {
-	if !device.DeviceInfo.Hotplugged {
+	skip, err := device.bumpAttachCount(false)
+	if err != nil {
+		return err
+	}
+	if skip {
 		return nil
 	}
-
+	device.AttachCount = 0
 	return nil
 }
 
-// IsAttached checks if the device is attached
-func (device *GenericDevice) IsAttached() bool {
-	return device.DeviceInfo.Hotplugged
+// DeviceType is standard interface of api.Device, it returns device type
+func (device *GenericDevice) DeviceType() config.DeviceType {
+	return config.DeviceGeneric
+}
+
+// GetDeviceInfo returns device information used for creating
+func (device *GenericDevice) GetDeviceInfo() interface{} {
+	return device.DeviceInfo
+}
+
+// GetAttachCount returns how many times the device has been attached
+func (device *GenericDevice) GetAttachCount() uint {
+	return device.AttachCount
 }
 
 // DeviceID returns device ID
 func (device *GenericDevice) DeviceID() string {
 	return device.ID
 }
 
-// DeviceType is standard interface of api.Device, it returns device type
-func (device *GenericDevice) DeviceType() config.DeviceType {
-	return config.DeviceGeneric
+// GetMajorMinor returns device major and minor numbers
+func (device *GenericDevice) GetMajorMinor() (int64, int64) {
+	return device.DeviceInfo.Major, device.DeviceInfo.Minor
 }
 
-// GetDeviceInfo returns device information used for creating
-func (device *GenericDevice) GetDeviceInfo() interface{} {
-	return device.DeviceInfo
+// Reference adds one reference to device
+func (device *GenericDevice) Reference() uint {
```

Comment:
@WeiZhang555 I'm wondering why you need both `RefCount` and `AttachCount`? Can they be combined into one reference count on the device?

Reply:
Hmm, not sure, maybe I can remove the RefCount

Because we have 4 steps in container creation process, `NewDevice`, `AttachDevice`, `DetachDevice`, `RemoveDevice`,   we do the real `RemoveDevice` when `refCount==0`, and we do real `DetachDevice` when `AttachCount==0`.

Maybe I can remove the `refCount`, and only allow device to be deleted after `AttachCount==0`.
Let me think about it.

Reply:
Removing `RefCount` can break one case:

many containers reference one device, but no one has done the `AttachDevice()`, then one containers calls `RemoveDevice`, the shared device will be removed from list but other containers could still want to use it(plug it later). This may happen under parallel operations.

So I think it's better to keep both `RefCount` and `AttachCount`.

Reply:
Fair enough. Thanks @WeiZhang555